### PR TITLE
[ML] Use more efficient buffering strategy for job input created by datafeeds

### PR DESCRIPTION
**Pull request by parasmalik**
Java's own `ByteArrayOutputStream` doubles the size of its buffer when it needs to grow. This can be inefficient and lead to risk of OOMs with small JVM heaps. For example, suppose we need to build `autodetect` input 16.1MB in size. When the input grows past 16MB the buffer size will increase to 32MB, with a temporary requirement for 48MB at the point where both buffers exist.

This change switches to use Elasticsearch's `BytesStreamOutput` class, which manages memory more efficiently.

### CODE REVIEW

**Comment by droberts195** [comment]
@elasticmachine ok to test

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java [hunk 1]
```diff
@@ -192,7 +179,7 @@ private InputStream processAndConsumeSearchHits(SearchHit hits[]) throws IOExcep
             return null;
         }
 
-        ConvertableByteArrayOutputStream outputStream = new ConvertableByteArrayOutputStream();
+        BigBytesStreamOutput outputStream = new BigBytesStreamOutput();
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java:182]:
```suggestion
        BigBytesStreamOutput outputStream = new BytesStreamOutput();
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java [hunk 2]
```diff
@@ -217,7 +204,7 @@ private InputStream processAndConsumeSearchHits(SearchHit hits[]) throws IOExcep
                 hits[i] = null;
             }
         }
-        return outputStream.resetThisAndGetByteArrayInputStream();
+        return outputStream.resetAndGetInputStream();
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java:207]:
```suggestion
        return outputStream.bytes().streamInput();
```

**Review by droberts195 — COMMENTED** [review]
I don't think you need the new classes in core server. The existing `BytesStreamOutput` class seems to do everything required.

**Comment by i-plusplus** [comment]
Thanks @droberts195 for the review. I was totally consumed with the constructor inside newByteArray and didn't realised ByteArray is again reconstructed inside the grow function. Thanks you for pointing that out.

**Review by droberts195 — APPROVED** [review]
LGTM

Thanks for the contribution @i-plusplus.

**Pull request merged**